File: 376-wiggle-subsequence/376-wiggle-subsequence.java
class Solution {
    public int wiggleMaxLength(int[] nums) {
        int n = nums.length;
        if(n <= 1) return n;
        int dp[] = new int[n];
        Arrays.fill(dp,1);
        int ans = 1;
        int k = 0;
        while(k < nums.length - 1 && nums[k] == nums[k+1]) k++;
        if(k == nums.length - 1) return 1;
        
        if(nums[k+1] - nums[k] > 0){
            for(int i=k+1;i<n;i++){
                for(int j=0;j<i;j++){
                    if(dp[j] % 2 == 0){
                        if(nums[j] > nums[i]){
                            dp[i] = Math.max(dp[i],dp[j] + 1);
                        }
                    }
                    else {
                        if(nums[j] < nums[i]){
                            dp[i] = Math.max(dp[i],dp[j] + 1);
                        }
                    }
                }
                ans = Math.max(ans,dp[i]);
            }
            return ans;
        }
        else {
            for(int i=k+1;i<n;i++){
                for(int j=0;j<i;j++){
                    if(dp[j] % 2 == 0){
                        if(nums[j] < nums[i]){
                            dp[i] = Math.max(dp[i],dp[j] + 1);
                        }
                    }
                    else {
                        if(nums[j] > nums[i]){
                            dp[i] = Math.max(dp[i],dp[j] + 1);
                        }
                    }
                }
                ans = Math.max(ans,dp[i]);
            }
            return ans;
        }
    }
}

File: 376-wiggle-subsequence/376-wiggle-subsequence.cpp
class Solution {
public:
    int wiggleMaxLength(vector<int>& nums) {
        
        int n = nums.size();
        if(n == 0) return 0;
        if(n == 1) return 1;
        
        vector<int> negative(n,1),positive(n,1);
        int ans = 1;
        for(int i=1;i<n;i++) {
            for(int j=0;j<i;j++) {
                int val = nums[i] - nums[j];
                
                if(val > 0) {
                    positive[i] = max(positive[i],negative[j] + 1);
                }
                else if(val < 0){
                    negative[i] = max(negative[i],positive[j] + 1);
                }
            }
            // cout << positive[i] << " " << negative[i] << endl;
            ans = max({ans,positive[i],negative[i]});
        }
        return ans;
    }
};